QAT v. 0.1 (#1230)
microsoft/qsharp-compiler: src/Passes/docs/src/UserGuide/BuildingLibrary.md

## Prerequisites

It is necessary for you to have the following tools before you get started on making a new build from this project:

- C++ compiler
- CMake

If you are compiling and using the command-line tool, these are necessary tools regardless of whether you plan to develop the library itself. The build configuration is handled by CMake, which informs you if a compiler needs to be installed.

> ### On OS X

Reviewer: ```suggestion
### On macOS
```

Author: Fixed

Those using OS X can install CMake through the use of a command-line tool called `brew`, as follows:

```sh
brew install cmake
```

> It should not even be necessary to install anything else on top of OS X, since the compiler comes preinstalled. There is no need to read the next subsection ([Developer prerequisites](#developer-prerequisites)) if you are not intending to modify the core library and can safely jump to the section [Library Dependencies](#library-dependencies)

Reviewer: ```suggestion
It should not even be necessary to install anything else on top of macOS, since the compiler comes preinstalled. There is no need to read the next subsection ([Developer prerequisites](#developer-prerequisites)) if you are not intending to modify the core library and can safely jump to the section [Library Dependencies](#library-dependencies)
```

Author: Updated.

### On Ubuntu 20.04

Additionally to CMake, you will also require Clang in order to build Ubuntu 20.04. You can accomplish this by following the steps below:

```sh
apt install clang-11 cmake
```

You can skip straight to [Library Dependencies](#library-dependencies) if you do not intend to work on developing the core library.

> 

Reviewer 2: Is Win supported?

Author: At this point it is not. I have not been given access to a Windows machine yet.

## Developer prerequisites

Development of the core library requires additional tools. These are used formatting, linting and managing code quality. For this purpose, the following tools and packages are used:

- Python 3
- Python packages specified in `requirements.txt`
- clang-format

> - clang-tidy

Reviewer: It may be helpful to provide a `dev-environment.yml` for `conda` that depends on all four of these items.

Author: I am not familiar with conda - is that the Windows package manager?

Reviewer: Some of your build scripts had called out to `conda`, so I was under the impression that was in-scope, sorry. Regardless, `conda` is a multi-language cross-platform package manager used to create virtual environments containing scientific development tooling. In this case, for example, an environment file like the one below could provide the above prerequisites cross-platform.

```yaml
# dev-environment.yml
name: qat-dev
channels:
- conda-forge
dependencies:
- python>=3.6
- click # Anything else from requirements.txt could go here, too,
        # if there's a matching conda package.
- clang
- clang-tools # Provides clang-format and clang-tidy
- pip
- pip:
  - ... # Anything from requirements.txt that doesn't have a
        # matching conda package can go here.
```

(As for Windows-specific package management, you may be thinking of `choco` or `winget`?)

Author: Thanks for the explanation and for the configuration. I am happy to add this.

Yes, it conda was used in the build script by suggestion of @idavis as a way to resolve issues getting LLVM and CMake to play nicely together. I did try `choco` for some other dependencies, but in the end, we never managed to get things to work. It is on my todo list as well.

Author: Added

The installation process varies depending on the platform you use. The following subsections provide details on how to install these tools on each platform.

> ### On OS X

Reviewer: ```suggestion
### On macOS
```

Author: Fixed.

TODO:

### On Ubuntu 20.04

Installing the clang tools on Ubuntu along with Python can be accomplished by running these commands:

```sh
apt install clang-format-11 clang-tidy-11
apt install python3 python3-pip
```

We recommend that you use version 11 of clang in order to be consistent with the version of LLVM on which the library depends. In general, the code should work with any version of clang. Make sure that the compiler environment variables are defined correctly in order to select the correct version:

```sh
export CC=clang-11
export CXX=clang++-11
```

### Common to all platforms

Last but not least, we install the Python libraries that are required:

```sh
      pip install -r requirements.txt

>       chmod +x manage

Reviewer: `chmod` isn't used on Windows, but this step is listed as being "common to all platforms."

Author: This would need to be revised once there is support for Windows. There are a number of places where the current setup is unlikely to work on this platform as it stands.

```

As a result, the `manage` tool will be available for you to help you make sure your code is high quality. Getting a contribution merged into the code base requires the manage tools CI flow to be successful. Thus, if you plan to contribute you will most likely need to complete this step.

## Library Dependencies

The QAT library itself is written in C++ and depends on LLVM for compilation and Google Test for testing purposes. Google test is checked out as a submodule whereas LLVM needs to be installed on the system. To get check out all submodules run

```sh
git submodule update --init --recursive
```

> ### Installing LLVM on OS X

Reviewer: ```suggestion
### Installing LLVM on macOS
```

Author: Fixed.

TODO:

### Installing LLVM on Ubuntu 20.04

For the installation of LLVM from Ubuntu we use the package manager `apt`. For full run installation, run the following command:

```sh
apt install  llvm-11 lldb-11 llvm-11-dev libllvm11 llvm-11-runtime
```

By doing this, we will ensure that CMake can find LLVM and that all headers and libraries used by the QAT library are available.

# Build instructions for users

First

```sh
cd Debug

> Cmake ..

Reviewer 2: > Cmake ..

I suspect you meant 
either `CMake` (the first 2 letters are capital)
or `cmake` (no capital letters)
here.

For now I have an impression that the commad `Cmake` will not be found on the case-sensitive file systems.

Author: Fixed - thanks .

> Cmake ..

Reviewer 2: On Win:
```
CMake Error at CMakeLists.txt:15 (find_package):
  Could not find a package configuration file provided by "LLVM" with any of
  the following names:

    LLVMConfig.cmake
    llvm-config.cmake

  Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
  "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
  provides a separate development package or SDK, be sure it has been
  installed.
-- Configuring incomplete, errors occurred!
```
I failed to find any of those 2 files on my machine in the following dirs (e.g. `C:\Program Files\LLVM11.1.0`):
```cmd
where clang
C:\Program Files\LLVM11.1.0\bin\clang.exe
C:\Program Files\LLVM12.0.1\bin\clang.exe
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\Llvm\x64\bin\clang.exe
```

Author: Unfortunately, I have not had the chance to try this on Windows yet so this answer may be incomplete. 

The package that is searched for is the source and runtime for `LLVM`. Hence, you will need to have these installed. They are not included with Clang and it is my understanding that you will need to compile these dependencies from scratch.
